pennylane/debugging.py

```diff
@@ -105,3 +108,75 @@ def get_snapshots(*args, **kwargs):
         return dbg.snapshots
 
     return get_snapshots
+
+
+class PLDB(pdb.Pdb):
+    """Custom debugging class integrated with Pdb."""
+
+    __active_dev = []
+
+    def __init__(self, *args, **kwargs):
+        """Initialize the debugger, and set custom prompt string."""
+        super().__init__(*args, **kwargs)
+        self.prompt = "[pldb]: "
+
+    @classmethod
+    def valid_context(cls):
+        """Determine if the debugger is called in a valid context.
+
+        Raises:
+            TypeError: Can't call breakpoint outside of a qnode execution
+            TypeError: Device not supported with breakpoint
+        """
+
+        if not qml.queuing.QueuingManager.recording() or not cls.is_active_dev():
+            raise TypeError("Can't call breakpoint outside of a qnode execution")
+
+        if cls.get_active_device().name not in ("default.qubit", "lightning.qubit"):
+            raise TypeError("Device not supported with breakpoint")
+
+    @classmethod
+    def add_device(cls, dev):
+        """Add a device to the global active device list.
+
+        Args:
+            dev (Union[Device, "qml.devices.Device"]): The active device
+        """
+        cls.__active_dev.append(dev)
```

Comment:
Yes exactly! In fact, instead of appending, we could probably just override the 1st entry in the list each time, since there shouldn't be more than one "active device" at a time.

Reply:
Yea, if only 1 device is a constraint, adapting to match that design sounds reasonable.

Reply:
If we only need to support one device, it could be worth making the active device just a `ClassVar`. It's minimal syntax and simple to use. If, however, we think in the future it is likely that we'll need to support multiple devices, then defining these methods makes sense.

Reply:
I don't think we will want multi-device support in the future,